src/components/QueryEditor/VisualQueryEditor/utils/utils.ts

```diff
@@ -220,3 +222,52 @@ export function filterColumns(
       tableSchema?.filter((c) => expression?.columns?.includes(toColumnName(c)))
     : tableSchema;
 }
+
+// return columns in use by an expression. If none has been specified, return the first
+// time and number columns from the table definition
+export function defaultTimeSeriesColumns(expression: QueryExpression, tableColumns: AdxColumnSchema[]): string[] {
+  const res: string[] = [];
+  if (expression.where.expressions?.length) {
+    (expression.where.expressions as QueryEditorArrayExpression[]).forEach((exp) => {
+      if (exp.expressions.length) {
+        (exp.expressions as QueryEditorOperatorExpression[]).forEach((e) => {
+          if (!res.includes(e.property.name)) {
+            res.push(e.property.name);
+          }
+        });
+      }
+    });
+  }
+  if (expression.reduce.expressions?.length) {
+    (expression.reduce.expressions as QueryEditorReduceExpression[]).forEach((exp) => {
+      if (!res.includes(exp.property.name)) {
+        res.push(exp.property.name);
+      }
+    });
+  }
+  if (expression.groupBy.expressions?.length) {
+    (expression.groupBy.expressions as QueryEditorGroupByExpression[]).forEach((exp) => {
+      if (!res.includes(exp.property.name)) {
+        res.push(exp.property.name);
+      }
+    });
+  }
+
+  const timeCols = tableColumns
+    .filter((cc) => toPropertyType(cc.CslType) === QueryEditorPropertyType.DateTime)
+    .map((c) => c.Name);
```

Comment:
```suggestion
  const timeCols = tableColumns
    .reduce<string[]>((cols, option) => {
      if (toPropertyType(option.CslType) === QueryEditorPropertyType.DateTime) {
        cols.push(option.Name);
      }
      return cols;
    }, [])
```
Maybe use `reduce` here and below to avoid looping twice?

Reply:
usually I don't like `reduce` loops that much because they are more difficult to read (and usually arrays are not that big to make a difference) but since here I am doing the same twice (4 loops in total), it's worth it :+1: